### ISSUE TITLE
sending parallel request produces error: 401 Unauthorized

### ISSUE THREAD

**Issue by AngelKyriako** [ging/licode#362]
We are creating a virtual class application using licode and I am against an issue during implementing the unit tests for the multimedia routes based on licode.

Assuming I have written a function (createRoomFunc) that creates a licode room through the nuve client library.

The following procedure that sends each request in series will work, since it requests the server again after a request has sent its corresponding response:
```
    describe('licode API create', function () {   
      it('should create a few rooms successfully', function (done) {   
        async.series([   
          function (next) {   
            createRoomFunc('testCreate', next);   
          },   
          function (next) {    
            createRoomFunc('testCreate2', next);   
          },   
          function (next) {   
            createRoomFunc('testCreate3', next);   
          },   
          function (next) {   
            createRoomFunc('testCreate4', next);
          },
          function (next) {
            createRoomFunc('testCreate5', next);
          }
        ], done);
      });
    });
```

However if the same code snippet is used with async.parallel it will crash sending back a 401 error message.

Any feedback on this aspect ?
I will keep the issue updated as I work around it

**Comment by zoopnin** [comment]
how to run the licode server? can u help me?

**Comment by AngelKyriako** [comment]
doesn't that work for u ?

http://lynckia.com/licode/install.html

2015-10-21 6:20 GMT+03:00 zoopnin <notifications@github.com>:

**Comment by pybean** [comment]
AFAIK nuve requires every request for the same service ID has to be sent and received in time order, in order to prevent the race condition.

If you see messages "Invalid timestamp or cnonce" from your nuve log, then that's the case.

My suggestion is to implement a FIFO queue on the server side to digest parallel requests from the client side, so the those parallel requests will be processed one by one.

**Comment by AngelKyriako** [comment]
Yeah I have set it up with a FIFO redis queue for the time being.

Currently I have one seperate queue for each seperate nuve method. 

I guess I should push every nuve method in the same queue to be sure that the 401 error is not thrown. Correct ?

At first I though this error occurs only for create(room, token)  & delete room requests but it seems it is thrown also for retrieve requests. Is there a reason the for retrieve request to be limited by the time order as well ?

Finally, are there any plans to setup a FIFO queue embedded nuve to prevent this ?

Thank you in advance

**Comment by lodoyun** [comment]
@pybean is right, that is a security mechanism to avoid replay attacks. Basically, you should try to send messages in order and avoiding bursts. Implementing a buffer or queue in Nuve would defeat the whole purpose of the mechanism.

**Comment by AngelKyriako** [comment]
Don't you think that this causes a performance issue for a good number of concurrent users on a website that uses licode as its webrtc library.

Wouldn't it be easier to have optional parameters to enable or disable the mechanism.
To use queues to avoid bursts is of course a good thing, but being mandatory seems a bit rough.
For some websites that extra validation layer could be useful while for others not.
How a website would like to handle concurrent request should be configured by itself, don't you think ?

Furthermore the message: '401 Unauthorized' is not descriptive enough for this scenario. In a future update it would be helpful for devs to receive something like 'concurrent requests are not allowed' or whatever else makes sense.

Finally, is I there a reason to use this security mechanism not only for write but for read requests as well ?

**Comment by kekkokk** [comment]
Do you have some news?
I'm new to Licode and i'm facing the same issue. Can I disable it in the config?

**Comment by AngelKyriako** [comment]
@kekkokk last time I checked there was not a way to disable this. Try implementing this with a redis queue.

**Comment by ryanhope2** [comment]
Apart from losing protection from replay attacks, are there other bad consequences to disabling the timestamp checking?  E.g. is it preventing a race condition that we know of?  or is it really for replay protection only?

**Comment by jcague** [comment]
it's for replay protection only, we're considering a better way to do it while adding support for multiple endpoints.

**Comment by ryanhope2** [comment]
We're considering submitting a patch to optionally disable it.  Would love to have a better way to prevent replay attack, but in our case that's less important due to our infrastructure than failing under moderate concurrency.
